doc/source/reference/arrays.promotion.rst

```diff
@@ -0,0 +1,263 @@
+operations involving a NumPy dtype and a Python scalar (``int``, ``float``,
+or ``complex``) *can* lose precision. For instance, it is probably intuitive
+that the result of an operation between a Python integer and a NumPy integer
+should be a NumPy integer. However, Python integers have arbitrary precision
+whereas all NumPy dtypes have fixed precision, so the arbitrary precision
+of Python integers cannot be preserved.
+
+More generally, NumPy considers the "kind" of Python scalars, but ignores
+their precision when determining the result dtype. This is often convenient.
+For instance, when working with arrays of a low precision dtype, it is usually
+desirable for simple operations with Python scalars to preserve the dtype.
+
+  >>> arr_float32 = np.array([1, 2.5, 2.1], dtype="float32")
+  >>> arr_float32 + 10.0  # undesirable to promote to float64
+  array([11. , 12.5, 12.1], dtype=float32)
+  >>> arr_int16 = np.array([3, 5, 7], dtype="int16")
+  >>> arr_int16 + 10  # undesirable to promote to int64
+  array([13, 15, 17], dtype=int16)
+
+In both cases, the result precision is dictated by the NumPy dtype.
+Because of this, ``arr_float32 + 3.0`` behaves the same as
+``arr_float32 + np.float32(3.0)``, and ``arr_int16 + 10`` behaves as
+``arr_int16 + np.int16(10.)``.
+
+As another example, when mixing NumPy integers with a Python ``float``
+or ``complex``, the result always has type ``float64`` or ``complex128``:
+
+  >> np.int16(1) + 1.0
+  np.float64(2.0)
+
+However, these rules can also lead to surprising behaviors when working with
+low precision dtypes.
+
+First, since the Python value is converted to a NumPy one before the operation
+can by performed, operations can fail with an error when the result seems
+obvious. For instance, ``np.int8(1) + 1000`` cannot continue because ``1000``
+exceeds the maximum value of an ``int8``. When the Python scalar
+cannot be coerced to the NumPy dtype, an error is raised:
+
+  >>> np.int8(1) + 1000
+  Traceback (most recent call last):
+    ...
+  OverflowError: Python integer 1000 out of bounds for int8
+  >>> np.int64(1) * 10**100
+  Traceback (most recent call last):
+  ...
+  OverflowError: Python int too large to convert to C long
+  >>> np.float32(1) + 1e300
+  np.float32(inf)
+  ... RuntimeWarning: overflow encountered in cast
+
+Second, since the Python float or integer precision is always ignored, a low
+precision NumPy scalar will keep using its lower precision unless explicitly
+converted to a higher precision NumPy dtype or Python scalar (e.g. via ``int()``,
+``float()``, or ``scalar.item()``). This lower precision may be detrimental to
+some calculations or lead to incorrect results, especially in the case of integer
+overflows:
+
+  >>> np.int8(100) + 100  # the result exceeds the capacity of int8
+  np.int8(-56)
+  ... RuntimeWarning: overflow encountered in scalar add
+
+Note that NumPy warns when overflows occur for scalars, but not for arrays;
```

Comment:
Hmm, that's actually surprising in itself, since we were thinking of making array scalars and arrays more similar...